File: python/find-sessions.py
#!/usr/bin/env python3
# find-sessions.sh — portiert nach python
# Quelle: shell, OpenClaw@gateway1:skills/tmux/scripts/find-sessions.sh
# auch in: OpenClaw@gateway2:skills/tmux/scripts/find-sessions.sh
# Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import argparse
import os
import subprocess
import sys
from pathlib import Path

def usage():
    print("""Usage: find-sessions.sh [-L socket-name|-S socket-path|-A] [-q pattern]

List tmux sessions on a socket (default tmux socket if none provided).

Options:
  -L, --socket       tmux socket name (passed to tmux -L)
  -S, --socket-path  tmux socket path (passed to tmux -S)
  -A, --all          scan all sockets under CLAWDBOT_TMUX_SOCKET_DIR
  -q, --query        case-insensitive substring to filter session names
  -h, --help         show this help""")

def list_sessions(label, tmux_cmd, query):
    try:
        result = subprocess.run(
            tmux_cmd + ["list-sessions", "-F", "#{session_name}\t#{session_attached}\t#{session_created_string}"],
            capture_output=True,
            text=True,
            check=True
        )
        sessions = result.stdout.strip()
    except subprocess.CalledProcessError:
        print(f"No tmux server found on {label}", file=sys.stderr)
        return 1

    if query:
        filtered_sessions = []
        for line in sessions.split('\n'):
            if query.lower() in line.lower():
                filtered_sessions.append(line)
        sessions = '\n'.join(filtered_sessions)

    if not sessions:
        print(f"No sessions found on {label}")
        return 0

    print(f"Sessions on {label}:")
    for line in sessions.split('\n'):
        if line.strip():
            parts = line.split('\t')
            if len(parts) >= 3:
                name, attached, created = parts[0], parts[1], parts[2]
                attached_label = "attached" if attached == "1" else "detached"
                print(f"  - {name} ({attached_label}, started {created})")
    return 0

def main():
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('-L', '--socket', dest='socket_name')
    parser.add_argument('-S', '--socket-path', dest='socket_path')
    parser.add_argument('-A', '--all', action='store_true', dest='scan_all')
    parser.add_argument('-q', '--query', dest='query')
    parser.add_argument('-h', '--help', action='store_true', dest='show_help')

    args, unknown = parser.parse_known_args()

    if args.show_help:
        usage()
        sys.exit(0)

    if unknown:
        print(f"Unknown option: {unknown[0]}", file=sys.stderr)
        usage()
        sys.exit(1)

    socket_name = args.socket_name or ""
    socket_path = args.socket_path or ""
    query = args.query or ""
    scan_all = args.scan_all

    socket_dir = os.environ.get('CLAWDBOT_TMUX_SOCKET_DIR', 
                                os.path.join(os.environ.get('TMPDIR', '/tmp'), 'clawdbot-tmux-sockets'))

    if scan_all and (socket_name or socket_path):
        print("Cannot combine --all with -L or -S", file=sys.stderr)
        sys.exit(1)

    if socket_name and socket_path:
        print("Use either -L or -S, not both", file=sys.stderr)
        sys.exit(1)

    if not subprocess.call(["which", "tmux"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) == 0:
        print("tmux not found in PATH", file=sys.stderr)
        sys.exit(1)

    if scan_all:
        socket_path_obj = Path(socket_dir)
        if not socket_path_obj.is_dir():
            print(f"Socket directory not found: {socket_dir}", file=sys.stderr)
            sys.exit(1)

        sockets = list(socket_path_obj.glob('*'))
        if not sockets:
            print(f"No sockets found under {socket_dir}", file=sys.stderr)
            sys.exit(1)

        exit_code = 0
        for sock in sockets:
            if sock.is_socket():
                result = list_sessions(f"socket path '{sock}'", ["tmux", "-S", str(sock)], query)
                if result != 0:
                    exit_code = result
        sys.exit(exit_code)

    tmux_cmd = ["tmux"]
    socket_label = "default socket"

    if socket_name:
        tmux_cmd.extend(["-L", socket_name])
        socket_label = f"socket name '{socket_name}'"
    elif socket_path:
        tmux_cmd.extend(["-S", socket_path])
        socket_label = f"socket path '{socket_path}'"

    list_sessions(socket_label, tmux_cmd, query)

if __name__ == "__main__":
    main()

File: javascript/find-sessions.js
#!/usr/bin/env node
// find-sessions.sh — portiert nach javascript
// Quelle: shell, OpenClaw@gateway1:skills/tmux/scripts/find-sessions.sh
// auch in: OpenClaw@gateway2:skills/tmux/scripts/find-sessions.sh
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

import { spawnSync } from 'child_process';
import { existsSync, statSync } from 'fs';
import { readdir } from 'fs/promises';
import { join } from 'path';

const usage = () => {
  console.error(`Usage: find-sessions.sh [-L socket-name|-S socket-path|-A] [-q pattern]

List tmux sessions on a socket (default tmux socket if none provided).

Options:
  -L, --socket       tmux socket name (passed to tmux -L)
  -S, --socket-path  tmux socket path (passed to tmux -S)
  -A, --all          scan all sockets under CLAWDBOT_TMUX_SOCKET_DIR
  -q, --query        case-insensitive substring to filter session names
  -h, --help         show this help`);
};

let socketName = '';
let socketPath = '';
let query = '';
let scanAll = false;
const socketDir = process.env.CLAWDBOT_TMUX_SOCKET_DIR || (process.env.TMPDIR || '/tmp') + '/clawdbot-tmux-sockets';

const args = process.argv.slice(2);
let i = 0;

while (i < args.length) {
  const arg = args[i];
  switch (arg) {
    case '-L':
    case '--socket':
      socketName = args[i + 1] || '';
      i += 2;
      break;
    case '-S':
    case '--socket-path':
      socketPath = args[i + 1] || '';
      i += 2;
      break;
    case '-A':
    case '--all':
      scanAll = true;
      i++;
      break;
    case '-q':
    case '--query':
      query = args[i + 1] || '';
      i += 2;
      break;
    case '-h':
    case '--help':
      usage();
      process.exit(0);
    default:
      console.error(`Unknown option: ${arg}`);
      usage();
      process.exit(1);
  }
}

if (scanAll && (socketName || socketPath)) {
  console.error('Cannot combine --all with -L or -S');
  process.exit(1);
}

if (socketName && socketPath) {
  console.error('Use either -L or -S, not both');
  process.exit(1);
}

const tmuxExists = spawnSync('which', ['tmux']).status === 0;
if (!tmuxExists) {
  console.error('tmux not found in PATH');
  process.exit(1);
}

const listSessions = (label, tmuxArgs) => {
  const tmuxCmd = ['tmux', ...tmuxArgs, 'list-sessions', '-F', '#{session_name}\t#{session_attached}\t#{session_created_string}'];
  
  const result = spawnSync(tmuxCmd[0], tmuxCmd.slice(1), { encoding: 'utf8' });
  
  if (result.status !== 0) {
    console.error(`No tmux server found on ${label}`);
    return false;
  }
  
  let sessions = result.stdout.trim();
  
  if (query) {
    const lines = sessions.split('\n').filter(line => line.toLowerCase().includes(query.toLowerCase()));
    sessions = lines.join('\n');
  }
  
  if (!sessions) {
    console.log(`No sessions found on ${label}`);
    return true;
  }
  
  console.log(`Sessions on ${label}:`);
  sessions.split('\n').forEach(line => {
    const [name, attached, created] = line.split('\t');
    const attachedLabel = attached === '1' ? 'attached' : 'detached';
    console.log(`  - ${name} (${attachedLabel}, started ${created})`);
  });
  
  return true;
};

if (scanAll) {
  if (!existsSync(socketDir) || !statSync(socketDir).isDirectory()) {
    console.error(`Socket directory not found: ${socketDir}`);
    process.exit(1);
  }
  
  let exitCode = 0;
  
  readdir(socketDir)
    .then(async files => {
      if (files.length === 0) {
        console.error(`No sockets found under ${socketDir}`);
        process.exit(1);
      }
      
      for (const file of files) {
        const fullPath = join(socketDir, file);
        if (statSync(fullPath).isSocket()) {
          const success = listSessions(`socket path '${fullPath}'`, ['-S', fullPath]);
          if (!success) exitCode = 1;
        }
      }
      
      process.exit(exitCode);
    })
    .catch(err => {
      console.error(err.message);
      process.exit(1);
    });
} else {
  const tmuxArgs = [];
  let socketLabel = 'default socket';
  
  if (socketName) {
    tmuxArgs.push('-L', socketName);
    socketLabel = `socket name '${socketName}'`;
  } else if (socketPath) {
    tmuxArgs.push('-S', socketPath);
    socketLabel = `socket path '${socketPath}'`;
  }
  
  listSessions(socketLabel, tmuxArgs);
}
